pmd-java/src/main/java/net/sourceforge/pmd/lang/java/typeresolution/ClassTypeResolver.java

```diff
@@ -404,19 +425,319 @@ public Object visit(ASTName node, Object data) {
         return super.visit(node, data);
     }
 
+    /**
+     * Searches a list of methods by trying the three phases of method overload resolution.
+     * https://docs.oracle.com/javase/specs/jls/se7/html/jls-15.html#jls-15.12.2
+     */
     public JavaTypeDefinition getBestMethodReturnType(List<MethodType> methods, ASTArgumentList arguments,
```

Comment:
all these new methods (and some old ones, such as `getLocalApplicableMethods`, `getApplicableMethods`, etc) are very specific to method reference resolution, and most of them are just internal helpers. From the `ClassTypResolver` we solely use a couple of them from our `visit()` methods. Move them to a helper class (final class, private constructor, all static methods) specific for Method reference resolution.

Also, that way we **could** write unit tests for such public methods, and be finer at testing this complex logic, even using the JSL samples as test scenarios